When audits have the same timestamp, only diff against older audits (by ID)

jameslovejoy opened the pull request:
We've run into an issue where the diffs aren't appearing correctly when there are multiple audits with the same timestamp.  

Because the condition currently checks `created_at <= ? AND id != ?`, it may compare the current audit with a newer one if the timestamps are equal, instead of comparing it with the previous audit as we're expecting.  I've added a test case for this scenario which hopefully explains the expected behavior more clearly than I can describe here.  By changing the condition to `created_at <= ? AND id < ?` we ensure that we're only comparing the current audit against an older state.

jameslovejoy committed:
Only diff against older audits

mguymon commented:
So two issues with this change is if the primary key is not sequential e.g. so if the newer version has a lower id number than the old version, or if the primary key is not a number, e.g. id is "abc". 

Not having sequential ids is probably an unlikely edge case, but it would be nice to support non-numeric ids.

Would it be possible to use the version column? Which is the undocumented new feature (my fault), `audit :name, version => true` with the example [spec](https://github.com/harleyttd/auditable/blob/master/spec/lib/version_spec.rb). Every audit gets a sequential version id, the latest will always be the highest version id. I added it to deal with a distributed async system, were the audit time stamps would not correctly reflect the version order.

mguymon commented:
I updated the README to show the possible configuration options are for audit

jameslovejoy commented:
Thanks for the information about versions.  I've switched our code to use it, and it's an improvement over just timestamps alone.  However, I'm still seeing an issue if two audits are created simultaneously (our updates are processed by multiple background workers).

To test simultaneous updates, I'm running two updates in two separate consoles that will fire at the same time:

``` ruby
c = Customer.first; trigger = "2014-06-18 00:00:00".to_datetime.to_i;
while(true); next unless Time.now.to_i > trigger; c.update_attribute(:first_name, "First"); break; end
```

``` ruby
c = Customer.first; trigger = "2014-06-18 00:00:00".to_datetime.to_i;
while(true); next unless Time.now.to_i > trigger; c.update_attribute(:last_name, "Last"); break; end
```

After the updates run, I see two audits saved with the same version number.  Based on the [comments in the code](https://github.com/harleyttd/auditable/blob/master/lib/auditable/auditing.rb#L259-L260), it sounds like there shouldn't be two audits with the same version number.  So, I've added a unique index and a retry to ensure a unique version number.  You can see the code changes here: https://github.com/SparkartGroupInc/auditable/compare/retry-on-record-not-unique

Let me know if you want me to open another pull request for those changes.  Unfortunately, I'm not sure how to write an automated test for this scenario with simultaneous updates.

mguymon commented:
Yeah, the uniqueness of the version is based on transaction support, but it looks like the transation are not happen sequentially. Can you see if adding pessimistic locking (`lock(true)`) solves the unique version issue?

```
audit.version = (audits.lock(true).maximum('version')||0) + 1
```

jameslovejoy commented:
After trying that, it works about 99% of the time, but I would occasionally run into: 

```
ActiveRecord::StatementInvalid: Mysql2::Error: Deadlock found when trying to get lock; try restarting transaction: INSERT INTO `audits`...
```

I tried moving the transaction to encapsulate the entire `save_audit` method and adding a lock to the `last_audit` lookup as well, but still ran into the error occasionally.

jameslovejoy commented:
Just FYI, we've deployed this branch in production, and it's working well for us: https://github.com/SparkartGroupInc/auditable/compare/retry-on-record-not-unique
